Modules/POSIX/Sources/POSIX/Regex.swift

```diff
@@ -0,0 +1,319 @@
+
+                // Use UTF8View for unicode regexes
+                let startIndexUTF8 = index(from: group.start, in: string.utf8)
+                let endIndexUTF8 = index(from: group.end, in: string.utf8)
+
+                let match = String(string.utf8[startIndexUTF8..<endIndexUTF8])!
+                groups.append(match)
+                groupIdx += 1
+            }
+
+            let indexOfEndOfMatchUTF8 = index(from: regexMatches[0].rm_eo, in: string.utf8)
+            let remainderString = String(string.utf8[indexOfEndOfMatchUTF8..<string.utf8.endIndex])!
+
+            guard remainderString.isEmpty else {
+                break
+            }
+            string = remainderString
+            
+        }
+
+        return groups
+    }
+
+    /// Given a string, replace substrings matching this regex with a template.
+    ///
+    /// - parameter template: String used to replace all substrings matching this Regex.
+    /// - parameter string:   String on which replacing takes place.
+    /// - parameter options:  Matching options like `firstCharacterNotAtBeginningOfLine` (`REG_NOTBOL` in POSIX parlance).
+    ///
+    /// - returns: New string in which all substrings matching the regex were replaced with the template.
+    public func replace(with template: String, in string: String, options: MatchOptions = []) -> String {
+        var string = string
+        let maxMatches = 10
+        var totalReplacedString: String = ""
+        let templateArray = [UInt8](template.utf8)
+
+        while true {
+            var regexMatches = [regmatch_t](repeating: regmatch_t(), count: maxMatches)
+            let result = regexec(&preg, string, regexMatches.count, &regexMatches, options.rawValue)
+
+            guard result == 0 else {
+                break // Unmatched regex
+            }
+
+            guard regexMatches[0].rm_eo != regexMatches[0].rm_so else {
+                break // matches the empty string: avoid infinite loop
+            }
+
+            let start = Int(regexMatches[0].rm_so)
+            let end   = Int(regexMatches[0].rm_eo)
+            var replacedStringArray = [UInt8](string.utf8)
+            replacedStringArray.replaceSubrange(start..<end, with: templateArray)
+            
+            guard let replacedString = String(bytes: replacedStringArray, encoding: .utf8) else {
+                break
+            }
+
+            let templateDelta = template.utf8.count - (end - start)
+            let offset = Int(end + templateDelta)
+            let templateDeltaIndex = replacedString.utf8.index(replacedString.utf8.startIndex, offsetBy: offset)
+
+            totalReplacedString += String(describing: replacedString.utf8[replacedString.utf8.startIndex ..< templateDeltaIndex])
+            let startIndex = string.utf8.index(string.utf8.startIndex, offsetBy: end)
+            string = String(describing: string.utf8[startIndex ..< string.utf8.endIndex])
```

Comment:
Hi, we cannot remove the `describing:` label, XCode insists that we should use it.

Regarding the use of `Foundation`, it is needed by 
```
guard let replacedString = String(bytes: replacedStringArray, encoding: .utf8) else {
    break
}
```

I do agree that these optimizations could be postponed to another PR :)